### DISCUSSION TITLE
Athens in Offline Mode

### DISCUSSION THREAD

**Discussion by marwan-at-work** [gomods/athens#1538] [General]
Should we add more configuration to enable Athens to work in offline mode? 
Should we add more configuration to have "partial responses. In other words, Athens is in "online mode" BUT if "go mod download" fails, then we just return the storage list and not the full list? 

Do you feel like Athens's Config file has gotten out of hand? Does it have too many configurations? Does that matter? 

Discuss!! 😋

**Comment by marwan-at-work** [comment]
One solution I can think of (to avoid more configuration) is that an Athens Admin can also create their own "mini GOPROXY" that will never error out, and just return an empty list if "go mod download" fails. Then he/she can point Athens to that GOPROXY for "go mod download" via the `GoBinaryEnvVars` 

But is that better or worse? Maybe users prefer more configuration over having to hand-write their own proxy?

**Comment by arschles** [comment]
Writing this from my phone right now, but I feel like there have been enough requests for some more controls, yes. What do you mean by “mini proxy”? Is this something they write and point Athens to?

**Comment by arschles** [comment]
Also I do feel like the config has too many fields, but that’s partially because we don’t have it documented that well (something I’m slowly working on). I haven’t recently researched how other similar systems do this but iirc Apache used to let you split up It’s config into multiple files

**Comment by marwan-at-work** [comment]
> What do you mean by “mini proxy”? Is this something they write and point Athens to?

Yes, exactly. I've actually suggested it in the issue here: https://github.com/gomods/athens/issues/1511#issuecomment-583580510

> Also I do feel like the config has too many fields, but that’s partially because we don’t have it documented that well (something I’m slowly working on). 

I actually disagree here. I feel like our config.dev.toml is super well documented. Every value has a blob of text describing what it does and why you might need it

Is there anything in particular in that config you find not well documented? 

But also, if you're talking about our docs-website, yeah I think we can potentially have the config.toml file render on the website so we don't have to double-document stuff.

**Comment by arschles** [comment]
>Yes, exactly. I've actually suggested it in the issue here: #1511 (comment)

I haven't dove into list caching much. So here's a ton of questions:

- If we did cache the `/list` endpoint, then you could theoretically pre-load an Athens and have it hermetically sealed, do I have that right?
- Would that be the thing you're calling a mini proxy?
- I wonder if we should literally create a new project just for a self-contained / hermetically sealed proxy. Seems like there's a clear line we can draw between Athens and the other thing. What do you think about that?

>Is there anything in particular in that config you find not well documented?

I was actually thinking that the config is not on the website and that's a problem. So, you actually covered that in this:

>But also, if you're talking about our docs-website, yeah I think we can potentially have the config.toml file render on the website so we don't have to double-document stuff.

Great minds think alike I guess 😄 . If we could render it on the site somehow that would be 🔥. I was also thinking that when discussions is public, we could use one or more discussions to explain some commonly (mis)used configs.

**Comment by arschles** [comment]
To add to this - I tried to come up with a solution for offline mode, and to give people better control over determinism in the `/list` endpoint (and `@latest`) in https://github.com/gomods/athens/issues/1532#issuecomment-586038155. 

I guess right now, this isn't a good place to discuss (see what I did there) it because there's nobody else here, but I thought I'd put it here anyway
